### ISSUE TITLE
pre-commit hook yields false positives

### ISSUE THREAD

**Issue by goodspark** [ansible/ansible-lint#753]
##### Summary

The ansible-lint hook always passes.

##### Issue Type

- Bug Report

##### Ansible and Ansible Lint details

```console (paste below)
ansible --version
ansible 2.9.7
  config file = /home/sam/tmp/anslint/ansible.cfg
  configured module search path = ['/home/sam/.ansible/plugins/modules', '/usr/share/ansible/plugins/modules']
  ansible python module location = /home/sam/tmp/anslint/venv/lib/python3.8/site-packages/ansible
  executable location = /home/sam/tmp/anslint/venv/bin/ansible
  python version = 3.8.1 (default, Jan 22 2020, 10:59:28) [GCC 7.4.0]

ansible-lint --version
ansible-lint 4.2.0
```

- ansible installation method: pip
- ansible-lint installation method: pip

##### OS / ENVIRONMENT
Ubuntu 18.04
git 2.26.2

##### STEPS TO REPRODUCE

I've attached a git archive of a repo as a minimal example. Extract, `git init`, add all files and commit, and then run both `ansible-lint` and `pre-commit run -a` to compare.

[example.tar.gz](https://github.com/ansible/ansible-lint/files/4560913/example.tar.gz)

##### Desired Behaviour

I expect failures for the asd role, since it's using the shell command unnecessarily.

When I run ansible-lint directly:
```console
ansible-lint -v
Found roles: 
Found playbooks: ansible/playbooks/roles/asd/tasks/main.yml ansible/playbooks/thing.yml
Examining ansible/playbooks/roles/asd/tasks/main.yml of type playbook
Examining ansible/playbooks/thing.yml of type playbook
Examining ansible/playbooks/roles/asd/tasks/main.yml of type tasks
[301] Commands should not change things if nothing needs doing
ansible/playbooks/roles/asd/tasks/main.yml:2
Task/Handler: asasd

[305] Use shell only when shell functionality is required
ansible/playbooks/roles/asd/tasks/main.yml:2
Task/Handler: asasd
```

##### Actual Behaviour

```console
pre-commit run -a
Ansible-lint.............................................................Passed
```

In addition, if you remove the shell command in the asd role and then add it back in a new commit (simulating a normal run of pre-commit), ansible-lint still yields a false positive.

```console
git diff --cached
diff --git a/ansible/playbooks/roles/asd/tasks/main.yml b/ansible/playbooks/roles/asd/tasks/main.yml
index ed97d53..fc4cf0c 100644
--- a/ansible/playbooks/roles/asd/tasks/main.yml
+++ b/ansible/playbooks/roles/asd/tasks/main.yml
@@ -1 +1,3 @@
 ---
+- name: asasd
+  shell: echo asd

git commit -m asd
Ansible-lint.............................................................Passed
[master ed1d4dc] asd
 1 file changed, 2 insertions(+)
```

**Comment by webknjaz** [comment]
@ssbarnea do you want to take a look?

**Comment by ssbarnea** [comment]
That is a side effect of the bug with auto detection, which is partially broken in 4.2.0, even on master. Once we fix this, I will try to reproduce it. For the moment I would say better to avoid 4.2.x

**Comment by weeheavy** [comment]
The exact same issue happens with ansible-lint v4.1.1a4 and the demo repo from OP:

```
(ansible-29) foo :: Downloads/ansible/playbooks ‹master*› » pre-commit run -a
Trim Trailing Whitespace.................................................Passed
Fix End of Files.........................................................Passed
Check for merge conflicts................................................Passed
Check for case conflicts.................................................Passed
Check Yaml...............................................................Passed
Detect AWS Credentials...................................................Passed
Detect Private Key.......................................................Passed
Ansible-lint.............................................................Passed
Detect platform READMEs needing attention.......(no files to check)Skipped
(ansible-29) foo :: Downloads/ansible/playbooks ‹master*› » ansible-lint
[301] Commands should not change things if nothing needs doing
roles/asd/tasks/main.yml:2
Task/Handler: asasd

[305] Use shell only when shell functionality is required
roles/asd/tasks/main.yml:2
Task/Handler: asasd
```

git version 2.26.2
macOS 19.4.0
```
(ansible-29) foo :: Downloads/ansible/playbooks ‹master*› » pip freeze | grep -E '(ansible|ansible-lint|pre-commit)=='
ansible==2.9.9
ansible-lint==4.2.0
pre-commit==2.5.1
```

.pre-commit-config.yml
```
- repo: https://github.com/ansible/ansible-lint.git
  # avoid 4.2.x right now: https://github.com/ansible/ansible-lint/issues/753#issuecomment-623418362
  rev: v4.1.1a4
  hooks:
    - id: ansible-lint
      files: \.(yaml|yml)$
```
Which version is currently working with pre-commit?

**Comment by ssbarnea** [comment]
Please run `pre-commit autoupdate` and run again, also double check that your hook definition matches the one in our examples (there where recent changes in the way we trigger it, mainly auto-detect).

PS. Current version is 4.3.x

**Comment by weeheavy** [comment]
Thanks for the hints.

`pre-commit autoupdate` updates from v4.1.1a4 -> v4.3.0a1 which is expected, but does not solve the issue, same thing happens. After double checking the hook definition, I found that [ansible-lint docs](https://docs.ansible.com/ansible-lint/configuring/configuring.html#pre-commit-setup) tell to use `sha` to specify the git tag, which returns a warning like `[WARNING] Unexpected key(s) present on https://github.com/ansible/ansible-lint.git: sha`. So my config above looks sane I think, except if I'm missing something else.

**Comment by weeheavy** [comment]
Update: https://github.com/ansible/ansible-lint/issues/649#issuecomment-568926113 fixes my issue. If that's how it should work, I think the docs need to be updated.

**Comment by ssbarnea** [comment]
These docs are more than an year old and refer to older version. The ones from inside the repository are correct. The job to update docs.ansible.com was somehow lost over time and we are working to assure docs are published on each merged change.
